File: Python/프로그래머스/Level2/행렬 테두리 회전하기.py
rows = 6
columns = 6
queries = [[2, 2, 5, 4], [3, 3, 6, 6], [5, 1, 6, 3]]


def solution(rows, columns, queries):
    answer = []
    board = [[] for i in range(rows)]  # 초기화
    for i in range(rows):
        for j in range(columns):
            board[i].append((j+1)+columns*i)

    for q in queries:
        top, left, bottom, right = q[0]-1, q[1]-1, q[2]-1, q[3]-1
        tmp = board[top][left]
        changed = []
        for y in range(top, bottom):  # 왼쪽
            board[y][left] = board[y+1][left]
            changed.append(board[y][left])

        for x in range(left, right):  # 아래
            board[bottom][x] = board[bottom][x+1]
            changed.append(board[bottom][x])

        for y in range(bottom, top, -1):  # 오른쪽
            board[y][right] = board[y-1][right]
            changed.append(board[y][right])

        for x in range(right, left, -1):  # 위
            board[top][x] = board[top][x-1]
            changed.append(board[top][x])
        board[top][left+1] = tmp
        changed.append(tmp)
        answer.append(min(changed))
        # print(changed)
    print(answer)
    return answer


solution(rows, columns, queries)

File: Javascript/프로그래머스/Level2/행렬 테두리 회전하기.js
function solution(rows, columns, queries) {
    var answer = [];
    let board = new Array(rows);
    for (let r = 0; r < rows; r++) {
        // board 초기화
        board[r] = new Array(columns);
        for (let c = 0; c < columns; c++) {
            board[r][c] = columns * r + (c + 1);
        }
    }

    for (let query of queries) {
        let [x1, y1, x2, y2] = query;
        x1 -= 1;
        y1 -= 1;
        x2 -= 1;
        y2 -= 1;
        let tmp = board[x1][y1]; // 구멍
        let changed = [];

        for (let r = x1; r < x2; r++) {
            // 왼쪽
            board[r][y1] = board[r + 1][y1];
            changed.push(board[r][y1]);
        }

        for (let c = y1; c < y2; c++) {
            //아래
            board[x2][c] = board[x2][c + 1];
            changed.push(board[x2][c]);
        }

        for (let r = x2; r > x1; r--) {
            // 오른쪽
            board[r][y2] = board[r - 1][y2];
            changed.push(board[r][y2]);
        }

        for (let c = y2; c > y1; c--) {
            //위
            board[x1][c] = board[x1][c - 1];
            changed.push(board[x1][c]);
        }
        board[x1][y1 + 1] = tmp;
        changed.push(board[x1][y1 + 1]);
        answer.push(Math.min.apply(null, changed));
    }

    return answer;
}
